src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -130,30 +131,65 @@ private static void ModifyUnmanagedDllResolver(Action<AssemblyLoadContext> resol
 
         private static IntPtr HostFxrResolver(Assembly assembly, string libraryName)
         {
-            var hostFxrLibName = "libhostfxr";
-            var libExtension = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "dylib" : "so";
+            string hostFxrLibName = "libhostfxr";
+            string libExtension = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "dylib" : "so";
 
-            if (!hostFxrLibName.Equals(libraryName))
+            // the DllImport hardcoded the name as hostfxr.
+            if (!hostFxrLibName.Equals(libraryName, StringComparison.OrdinalIgnoreCase) && !libraryName.Equals("hostfxr", StringComparison.OrdinalIgnoreCase))
             {
                 return IntPtr.Zero;
             }
 
-            var hostFxrRoot = Path.Combine(DotnetPath.Value, "host", "fxr");
+            string hostFxrRoot = Path.Combine(DotnetPath.Value, "host", "fxr");
             if (Directory.Exists(hostFxrRoot))
             {
+                var fileEnumerable = new FileSystemEnumerable<SemanticVersion?>(
+                    directory: hostFxrRoot,
+                    transform: static (ref FileSystemEntry entry) => SemanticVersionParser.TryParse(entry.FileName.ToString(), out var version) ? version : null)
+                {
+                    ShouldIncludePredicate = static (ref FileSystemEntry entry) => entry.IsDirectory
+                };
+
                 // Load hostfxr from the highest version, because it should be backward-compatible
-                var hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
-                            .Max(str => SemanticVersionParser.TryParse(str, out var version) ? version : null);
+                SemanticVersion? hostFxrVersion;
+                
+                try
+                {
+                    hostFxrVersion = fileEnumerable.Max();
+                }
+                catch (ArgumentException)
+                {
+                    // LINQ Max throws when the list is empty.
+                    Console.Error.WriteLine($"No child folder was found in '{hostFxrRoot}'.");
+                    return IntPtr.Zero;
+                }
```

Comment:
`Enumerable.Max` is documented to throw `InvalidOperationException` if the sequence is empty. Does not seem to apply to the overload being called here, though. `Max` just returns null: https://dotnetfiddle.net/JRnHee

```suggestion
                SemanticVersion? hostFxrVersion = fileEnumerable.Max();
```

Reply:
my bad, i misread the document. There are indeed different behavior for different variations of this method...